### PR TITLE
Migration guide 3.0.0

### PR DESCRIPTION
**Pull request by angrykoala**
# Description
Docs for migrating to 3.0 from 2.x

### CODE REVIEW

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:33]: nitpicking: couldn't this just be `schema`?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:52]: .. such as `Director` remain ..

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:146]: More of an accessibility note. The link to `@node directive` is very helpful but it wasn't very obvious to me that this is a link.

**Review by tbwiss — COMMENTED** [review]
Nice one! Commented on some minor things.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:33]: Yes :sweat_smile: , I decided to go with this as some people may not be completely familiar with object destructuring, but happy to change if you think it is cleaner that way

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:146]: What alternative could we use? just "@node directive" without code blocks?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:33]: Yes I see. Good point! Let's opt for the most familiar version and leave it as is.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:146]: The alternative without code blocks seems good. But on the other hand this may also just be some useless nitpicking from my side 😅 . We can leave it as is.

Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:146]: If you think it helps accessibility I'm happy to change it
```suggestion
please update your clients to use the new query names or use the `plural` option in the xref::type-definitions/database-mapping.adoc#_plural[@node directive]
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:35]: Possibly also worth documenting this using `.then`? I imagine this will be appropriate for a lot of users.

```
----
const neoSchema = new Neo4jGraphQL({ typeDefs, driver });
neoSchema.getSchema().then((schema) => {
  const server = new ApolloServer({
    schema: schema
  });
});
----
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:83]:
```suggestion
Queries using old relationship filters, will now need to use `{relationship}_SOME`. For example:
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:122]: This is more than just `jwksEndpoint`? I would suggest this should have two subsections, one for JWT users and one for JWKS users.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:156]:
```suggestion
Neo4j 4.1 is no longer supported in 3.0.0, inline with the [supported versions list](https://neo4j.com/developer/kb/neo4j-supported-versions/).
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:159]: A bit more boilerplate, but I prefer something like this:

```suggestion
`graphql@^15.0.0` is no longer supported, please upgrade to `graphql@^16.0.0` using `npm` or the package manager of your choice.
```

**Review by darrellwarde — COMMENTED** [review]
Something quite major which is missing is for users who use custom directives. Compare what is about to be published (https://neo4j-graphql-docs-977.surge.sh/graphql-manual/2.0/type-definitions/custom-directives/) to something like https://www.apollographql.com/docs/apollo-server/v2/schema/creating-directives/#implementing

I would say this is one of the biggest changes!

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (wheresrhys) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:54]: Are there any more details available of the effect this will have? e.g. will the server start to error if a rogue relationship is created in the DB? Can this functionality be turned off?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:35]: mmm, documenting both would be heavily redundant, so I guess it is promises vs async await

Thoughts?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (wheresrhys) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:46]: What will be the API behaviour if Movie does not have any relationships to Actors? I'm hoping this change doesn't mean there'll be an error.

Comment (darrellwarde) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:46]: Behaviour will be identical - this always returned an empty List, and this new type definition constraint matches the possible return values.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:35]: Does this come down to a matter a of (personal) preference? Then I vote for async await due to its slightly better readability IMO. But if the majority of our users are more familiar to promises then we should opt for that.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:54]: Hi @wheresrhys, we will try to be as comprehensive with potential pitfalls as possible. As you mention, there is a change the server may return an error on **mutations**  if your database has rogue relationships. Please keep in mind that if that is the case, the current behaviour is already undetermined and **will** cause problems, at least for now we do not plan on making this optional, as the current behaviour is actually buggy.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:35]: I agree with you, although I think most of our examples are already riddled with async/await. Thoughts on this @darrellwarde @danstarns ?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:127]: Agree, I think I'll wait until auth is merged to fix this, as we will probably want to link to the updated docs to avoid duplication

**Comment by darrellwarde** [comment]
The more I read this, the more I think the "Generated Types" page is just overkill and it should just be on a single page.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:13]:
```suggestion
== Asynchronous schema generation
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:15]:
```suggestion
as a `Promise`. This means that creating a server now requires awaiting for that method:
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:39]:
```suggestion
This release contains an overhaul of our relationship validations, which will require a few changes to the schema.
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:41]:
```suggestion
=== Many-to-* relationships
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:42]:
```suggestion
To improve consistency and validation, **all** "many-to-*" relationships need to be defined as _required_ in the schema:
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:52]:
```suggestion
Note that any other notation, such as `[Actor]` or `[Actor!]` will **not** be valid. "One-to-one" relationships
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:56]:
```suggestion
Runtime checks for "one-to*" relationships have been added in this release, ensuring that the correct number of relationships exist. This means that some
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:57]: Why would this have happened? It will be worthwhile giving some background as to why some databases might have ended up in this state.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:65]:
```suggestion
query {
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:74]:
```suggestion
query {
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:88]:
```suggestion
query {
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:103]:
```suggestion
query {
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:126]: Worth showing a before and after?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:146]:
```suggestion
https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets[JSON Web Key Sets] are now supported through `Neo4jGraphQLAuthJWKSPlugin`.
```

**Review by darrellwarde — CHANGES_REQUESTED** [review]

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:57]: I can think of manual changes to the DB, bugs, or changes in typedefs as the main culprits, anything else we should list there?

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:116]: mmm, I wasn't completely sure, but yeah, makes sense

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:185]:
```suggestion
=== Removal of nested operation fields for `connectOrCreate`
```

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:186]:
```suggestion
Input types for `onCreate` in `connectOrCreate` operations no longer accept relationship fields. They were originally added in error and did not function as one would expect, so there is no regression in functionality.
```

**Review by darrellwarde — CHANGES_REQUESTED** [review]

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (darrellwarde) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:57]: More like, due to the fact that we never enforced this before, it's likely that through normal use of our library before this version, your database may have ended up in this state.

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:60]: More a question here but do we want this to be a list with each element on a new line? If not, ignore this comment please. If so then it isn't the case just now: https://neo4j-graphql-docs-977.surge.sh/graphql-manual/3.0/guides/v3-migration/#_relationship_cardinality

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (tbwiss) [inline docs/modules/ROOT/pages/guides/v3-migration/index.adoc:214]: These two blocks seem to have a lot of similar content. Shall we combine them?

**Review by danstarns — APPROVED** [review]

File: docs/modules/ROOT/pages/guides/v3-migration/index.adoc
Comment (angrykoala) [reply docs/modules/ROOT/pages/guides/v3-migration/index.adoc:60]: Good catch. however, due to Darrell's comment this is not going to be a list, so problem solved XD

**Pull request merged**
